// @var3n1k/server-core/build/src/module/dataset/constants/default/date/week.js
// import * as GlobalModule from '../../@module.js'
export default {
    Day: {
        Name: {
            1: {
                Full: {
                    Single: { 1: `Понедельник`, 2: `Понедельника`, 3: `Понедельник`, 4: `Понедельнику`, 5: `Понедельнике`, 6: `Понедельником` },
                    Multiple: { 1: `Понедельники`, 2: `Понедельников`, 3: `Понедельники`, 4: `Понедельникам`, 5: `Понедельниках`, 6: `Понедельниками` },
                },
                Short: `Пн`,
            },
            2: {
                Full: {
                    Single: { 1: `Вторник`, 2: `Вторника`, 3: `Вторник`, 4: `Вторнику`, 5: `Вторнике`, 6: `Вторником` },
                    Multiple: { 1: `Вторники`, 2: `Вторников`, 3: `Вторники`, 4: `Вторникам`, 5: `Вторниках`, 6: `Вторниками` },
                },
                Short: `Вт`,
            },
            3: {
                Full: {
                    Single: { 1: `Среда`, 2: `Среды`, 3: `Среду`, 4: `Среде`, 5: `Среде`, 6: `Средой` },
                    Multiple: { 1: `Среды`, 2: `Сред`, 3: `Среды`, 4: `Средам`, 5: `Средах`, 6: `Средами` },
                },
                Short: `Ср`,
            },
            4: {
                Full: {
                    Single: { 1: `Четверг`, 2: `Четверга`, 3: `Четверг`, 4: `Четвергу`, 5: `Четверге`, 6: `Четвергом` },
                    Multiple: { 1: `Четверги`, 2: `Четвергов`, 3: `Четверги`, 4: `Четвергам`, 5: `Четвергах`, 6: `Четвергами` },
                },
                Short: `Чтв`,
            },
            5: {
                Full: {
                    Single: { 1: `Пятница`, 2: `Пятницы`, 3: `Пятницу`, 4: `Пятнице`, 5: `Пятнице`, 6: `Пятницей` },
                    Multiple: { 1: `Пятницы`, 2: `Пятниц`, 3: `Пятницы`, 4: `Пятницам`, 5: `Пятницах`, 6: `Пятницами` },
                },
                Short: `Птн`,
            },
            6: {
                Full: {
                    Single: { 1: `Суббота`, 2: `Субботы`, 3: `Субботу`, 4: `Субботе`, 5: `Субботе`, 6: `Субботой` },
                    Multiple: { 1: `Субботы`, 2: `Суббот`, 3: `Субботы`, 4: `Субботам`, 5: `Субботах`, 6: `Субботами` },
                },
                Short: `Сб`,
            },
            7: {
                Full: {
                    Single: { 1: `Воскресенье`, 2: `Воскресенья`, 3: `Воскресенье`, 4: `Воскресенью`, 5: `Воскресенье`, 6: `Воскресеньем` },
                    Multiple: { 1: `Воскресенья`, 2: `Воскресений`, 3: `Воскресенья`, 4: `Воскресеньям`, 5: `Воскресеньях`, 6: `Воскресеньями` },
                },
                Short: `Вскр`,
            },
        },
    },
};
