// @var3n1k/server-core/build/src/module/dependencies/default/node/__node.js
export { default as assert } from 'node:assert';
export * as async_hooks from 'node:async_hooks';
export * as buffer from 'node:buffer';
export * as child_process from 'node:child_process';
export * as cluster from 'node:cluster';
export { default as console } from 'node:console';
export { default as constants } from 'node:constants';
export * as crypto from 'node:crypto';
export * as dgram from 'node:dgram';
// export * as diagnostics_channel from 'node:diagnostics_channel'  // Deno | ?Bun
export * as dns from 'node:dns';
export * as domain from 'node:domain';
export { default as events } from 'node:events';
export { default as fs } from 'node:fs';
export * as http from 'node:http';
// export * as http2 from 'node:http2'  // Deno | ?Bun
export * as https from 'node:https';
// export { default as inspector } from 'node:inspector'  // Deno | ?Bun
// export { default as module } from 'node:module'  // Bun
export * as net from 'node:net';
export * as os from 'node:os';
export { default as path } from 'node:path';
export * as perf_hooks from 'node:perf_hooks';
export { default as process } from 'node:process';
// export { default as punycode } from 'node:punycode'  // Deno | ?Bun
export * as querystring from 'node:querystring';
export * as readline from 'node:readline';
export { default as stream } from 'node:stream';
export * as string_decoder from 'node:string_decoder';
export { default as timers } from 'node:timers';
export * as tls from 'node:tls';
export * as tty from 'node:tty';
export * as url from 'node:url';
export * as util from 'node:util';
export * as zlib from 'node:zlib';
