// @var3n1k/server-core/build/src/module/dataset/dictionary/symbol.js
// import * as GlobalModule from '../../@module.js'
export default {
    Single: {
        Service: {
            Separator: {
                Space: {
                    Default: ` `,
                    Unicode: `\u00A0`,
                },
            },
            Output: {
                Status: {
                    Success: { Colored: `✔` },
                    Error: { Default: `×`, Italic: `✘`, Colored: `✖` },
                    Warning: { Colored: `⚠` },
                },
            },
            Mark: {
                Status: {
                    Check: { On: `☑`, Off: `☐`, Disabled: `☒` },
                    Radio: { On: `◉`, Off: `◯` },
                    Dot: {
                        Center: { On: `•`, Off: `◦`, Small: { On: `·` } },
                    },
                    Star: { On: `★`, Off: `☆` },
                },
                Question: { Bold: `？`, Default: '?', Thin: `﹖` },
                Information: { Default: `i`, Colored: `ℹ` },
                MultiDot: { Bold: `…`, Thin: `⋯` },
            },
            // Arrow: {
            //     SingleSide: {
            //     },
            //     MultipleSide: {
            //         UpDown: {
            //             Single: {
            //                 Default: { Filled: undefined, Bare: `↕` },
            //                 Small: { Filled: undefined, Bare: undefined },
            //                 Large: { Filled: undefined, Bare: undefined },
            //             },
            //             Multiple: {
            //                 Double: {
            //                     Default: undefined,
            //                     Small: { Filled: undefined, Bare: undefined },
            //                     Large: { Filled: undefined, Bare: undefined },
            //                 },
            //             },
            //         },
            //     },
            // },
            // Pointer: {
            //     SingleSide: {
            //         Left: {
            //             Single: {
            //                 Default: { Filled: undefined, Bare: `>` },
            //                 Small: { Filled: `‣`, Bare: `›` },
            //                 Large: { Filled: `▸`, Bare: `❯` },
            //             },
            //             Multiple: {
            //                 Double: {
            //                     Default: undefined,
            //                     Small: { Filled: undefined, Bare: `»` },
            //                     Large: { Filled: undefined, Bare: undefined },
            //                 },
            //             },
            //         },
            //     },
            //     MultipleSide: {
            //     },
            // },
            // Line: {
            //     Center: {
            //         Large: { Bold: `─` },
            //         Small: { Thin: `－` },
            //     },
            // },
            Paragraph: {
                Default: `§`,
                Pilcrow: { 1: `¶`, 2: `❡` },
            },
        },
        Graphic: {
            Pencil: { UpRight: `✐`, Colored: `✏`, DownRight: `✎` },
        },
        Math: {
            Operator: { IdenticalTo: `≡`, PlusMinus: `±`, Obelus: `÷` },
        },
    },
    Collection: {
        Numbers: [`0`, `1`, `2`, `3`, `4`, `5`, `6`, `7`, `8`, `9`],
        Special: {
            All: ['`', `~`, `!`, `@`, `"`, `#`, `№`, `$`, `;`, `%`, `^`, `:`, `&`, `?`, `*`, `(`, `)`, `_`, `-`, `+`, `=`, `{`, `}`, `[`, `]`, `'`, `,`, `.`, `<`, `>`, `|`, `\\`, `/`],
            Group: {
                HashFriendly: [`~`, `!`, `@`, `#`, `$`, `&`, `*`, `(`, `)`, `_`, `-`, `{`, `}`, `[`, `]`, `|`, `\\`, `/`],
                PasswordFriendly: [`!`, `@`, `#`, `$`, `&`, `_`, `-`],
            },
        },
    },
};
