// @var3n1k/server-core/build/src/module/functions/parameter/type/manager/default/date/difference/GetValues.js
import * as GlobalModule from '../../../../../../../@module.js';
//* -------------------------------------------------------------------------
//* Realization
//* -------------------------------------------------------------------------
export default function GetValues(from, to) {
    const Validator = GlobalModule.Classes.Validator;
    Validator.Strict(from, new Validator().Default.Date().Required());
    Validator.Strict(to, new Validator().Default.Date().Required());
    const DifferenceAsDate = new Date(to.getTime() - from.getTime());
    const DifferenceAsDateValues = GlobalModule.Functions.Parameter.Type.Manager.Default.Date.GetValues(DifferenceAsDate, false);
    const DifferenceInMilliseconds = {
        Local: DifferenceAsDateValues.Milliseconds,
        Total: DifferenceAsDate.getTime(),
    };
    const DifferenceInSeconds = {
        Local: DifferenceAsDateValues.Seconds,
        Total: Math.floor(DifferenceInMilliseconds.Total / 1000),
    };
    const DifferenceInMinutes = {
        Local: DifferenceAsDateValues.Minutes,
        Total: Math.floor(DifferenceInSeconds.Total / 60),
    };
    const DifferenceInHours = {
        Local: DifferenceAsDateValues.Hours,
        Total: Math.floor(DifferenceInMinutes.Total / 60),
    };
    const DifferenceInDays = {
        Local: DifferenceAsDateValues.DayOfTheMonth - 1,
        Total: Math.floor(DifferenceInHours.Total / 24),
    };
    const DifferenceInWeeks = {
        Local: Math.floor(DifferenceInDays.Local / 7),
        Total: Math.floor(DifferenceInDays.Total / 7),
    };
    const DifferenceInMonths = {
        Local: DifferenceAsDateValues.Month.Number - 1,
        Total: Math.floor(DifferenceInDays.Total / 30),
    };
    const DifferenceInYears = {
        Local: DifferenceAsDateValues.Year.Full - 1970,
        Total: Math.floor(DifferenceInDays.Total / 365),
    };
    const DatesDifferenceValues = {
        Years: DifferenceInYears,
        Months: DifferenceInMonths,
        Weeks: DifferenceInWeeks,
        Days: DifferenceInDays,
        Hours: DifferenceInHours,
        Minutes: DifferenceInMinutes,
        Seconds: DifferenceInSeconds,
        Milliseconds: DifferenceInMilliseconds,
    };
    return DatesDifferenceValues;
}
