// @var3n1k/server-core/build/src/module/dataset/constants/custom/color/hex.js
// import * as GlobalModule from '../../@module.js'
const DefaultColors = {
    Black: `#000000`,
    White: `#FFFFFF`,
    Grey: {
        Default: {
            Initial: `#808080`,
            Dark: { 100: `#111111`, 90: `#222222`, 80: `#333333`, 70: `#444444`, 60: `#555555`, 50: `#666666` },
            Light: { 50: undefined, 40: undefined, 30: `#A9A9A9`, 20: undefined, 10: undefined, 0: undefined },
        },
        Custom: {},
    },
    Red: {
        Default: {
            Initial: `#FF0000`,
            Dark: { 100: undefined, 90: undefined, 80: undefined, 70: undefined, 60: undefined, 50: undefined },
            Light: { 50: `#FF1010`, 40: `#FF2020`, 30: `#FF3030`, 20: `#FF4040`, 10: `#FF4554`, 0: `#FF6060` },
        },
        Custom: {},
    },
    Yellow: {
        Default: {
            Initial: `#DDED26`,
            Dark: { 100: undefined, 90: undefined, 80: undefined, 70: undefined, 60: undefined, 50: undefined },
            Light: { 50: undefined, 40: undefined, 30: undefined, 20: undefined, 10: undefined, 0: undefined },
        },
        Custom: {},
    },
    Green: {
        Default: {
            Initial: `#3AFC1C`,
            Dark: { 100: undefined, 90: undefined, 80: undefined, 70: undefined, 60: undefined, 50: undefined },
            Light: { 50: undefined, 40: undefined, 30: undefined, 20: undefined, 10: undefined, 0: undefined },
        },
        Custom: {},
    },
    Blue: {
        Default: {
            Initial: `#1B81A8`,
            Dark: { 100: undefined, 90: undefined, 80: undefined, 70: undefined, 60: undefined, 50: undefined },
            Light: { 50: undefined, 40: undefined, 30: undefined, 20: undefined, 10: undefined, 0: undefined },
        },
        Custom: {},
    },
};
export default {
    Default: DefaultColors,
    Custom: {
        Console: {
            Element: {
                Default: {
                    Message: `#D0D0D0`,
                    Number: `#A6CFA8`, // #C09D00
                },
                Custom: {},
            },
            Log: {
                Preset: {
                    Success: {
                        Tag: `#3AFC1C`,
                    },
                    Info: {
                        Tag: `#34E2E2`,
                    },
                    Warning: {
                        Tag: `#FFBB00`,
                    },
                    Error: {
                        Tag: `#EF2929`,
                        Code: {
                            Name: `#EF2929`,
                        },
                    },
                },
            },
        },
    },
};
