// @var3n1k/server-core/build/src/module/dependencies/default/__default.js
export { default as assert } from 'assert';
export * as async_hooks from 'async_hooks';
export * as buffer from 'buffer';
export * as child_process from 'child_process';
// export * as cluster from 'cluster'  // Bun
export { default as console } from 'console';
export { default as constants } from 'constants';
export * as crypto from 'crypto';
// export * as dgram from 'dgram'  // Bun
// export * as diagnostics_channel from 'diagnostics_channel'  // Deno | ?Bun
export * as dns from 'dns';
export * as domain from 'domain';
export * as dotenv from 'dotenv';
export { default as events } from 'events';
export { default as fs } from 'fs';
export * as http from 'http';
// export * as http2 from 'http2'  // Deno | ?Bun
export * as https from 'https';
// export * as inspector from 'inspector'  // Deno | ?Bun
// export { default as module } from 'module'  // Bun
export * as net from 'net';
export * as os from 'os';
export { default as path } from 'path';
export * as perf_hooks from 'perf_hooks';
export { default as process } from 'process';
// export { default as punycode } from 'punycode'  // Deno | ?Bun
export * as querystring from 'querystring';
export * as readline from 'readline';
export { default as stream } from 'stream';
export * as string_decoder from 'string_decoder';
export { default as timers } from 'timers';
export * as tls from 'tls';
export * as tty from 'tty';
export * as url from 'url';
export * as util from 'util';
export * as zlib from 'zlib';
export * as Node from './node/__node.js';
export * as Testing from './testing/__testing.js';
